### PR TITLE
Fixing system type deserialization in classes

### PR DESCRIPTION
**Pull request by Ellerbach**
## Description

Fixing system type deserialization in classes

## Motivation and Context

- Fixing system type deserialization in classes
- Add exhaustive system type deserialization to avoid cast issues
- Fixes nanoframework/Home#891.

## How Has This Been Tested?

With unit tests and on a real device

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (work on Unit Tests, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by josesimoes — APPROVED** [review]
👌🏻

File: nanoFramework.Json/JsonConvert.cs [hunk 1]
```diff
@@ -442,23 +475,56 @@ private static object PopulateObject(JsonToken rootToken, Type rootType, string
...
+                                                memberValueArrayList.Add(Convert.ToInt64(value.Value.ToString()));
+                                            }
+                                            else if (memberPropGetMethod.ReturnType.Name.Contains("SByte"))
+                                            {
+                                                memberValueArrayList.Add(Convert.ToSByte(value.Value.ToString()));
+                                            }
                                             else if (memberPropGetMethod.ReturnType.Name.Contains("Byte"))
                                             {
-                                                memberValueArrayList.Add(Convert.ToByte(value.Value.ToString()));
+                                                memberValueArrayList.Add(Convert.ToSByte(value.Value.ToString()));
```
Comment (mboud) [inline nanoFramework.Json/JsonConvert.cs:508]: ToByte !

File: nanoFramework.Json/JsonConvert.cs [hunk 2]
```diff
@@ -1030,7 +1132,24 @@ private static JsonToken ParseValue(ref LexToken token)
...
+                    }
+                    catch
+                    {
+                        try
+                        {
+                            return new JsonValue(long.Parse(token.TValue));
+                        }
+                        catch
+                        {
+                            {
```
Comment (mboud) [inline nanoFramework.Json/JsonConvert.cs:1147]: Is the extra set of curly braces voluntary ?

File: nanoFramework.Json.Test/JsonUnitTests.cs [hunk 3]
```diff
@@ -846,6 +846,69 @@ public void SerializeObjectAsAProperty()
...
+        public void DeserializeArrayToDeserialize()
+        {
+            ArrayToDeserialize obj0 = new() { Prop1 = 1, Prop2 = "prop2", Prop3 = true, Prop4 = 67890123 };
+            ArrayToDeserialize obj1 = new() { Prop1 = -42, Prop2 = "second2", Prop3 = false, Prop4 = 123456 };
+            ArrayToDeserialize[] array = new[] { obj0, obj1 };
+            var json = JsonConvert.SerializeObject(array);
+
+            var deser = JsonConvert.DeserializeObject(json, typeof(ArrayToDeserialize[])) as ArrayToDeserialize[];
+
+            Assert.Equal(deser.Length ,array.Length, "Aray length");
```
Comment (mboud) [inline nanoFramework.Json.Test/JsonUnitTests.cs:878]: I don't know how picky you accept PR comments, but there's a typo : "Array length", and the first parameter should have the space after the comma.

File: nanoFramework.Json.Test/JsonUnitTests.cs [hunk 4]
```diff
@@ -846,6 +846,69 @@ public void SerializeObjectAsAProperty()
...
+            Assert.Equal((sbyte)-42, deser.OneSByte, "SByte");
+            Assert.Equal((short)1234, deser.OneInt16, "Int16");
+            Assert.Equal((ushort)5678, deser.OneUInt16, "UInt16");
+            Assert.Equal(-789012, deser.OneInt32, "Int32");
+            Assert.Equal((uint)78912, deser.OneUInt32, "UInt32");
+            Assert.Equal((long)-1234567, deser.OneInt64, "Int64");
+            Assert.Equal((ulong)1234567, deser.OneUInt64, "UInt64");
+            Assert.Equal((float)34.45, deser.OneSingle, "Single");
+            Assert.Equal((double)45678.23, deser.OneDouble, "Double");
+            Assert.True(deser.OneBoolean,"Boolean true");
```
Comment (mboud) [inline nanoFramework.Json.Test/JsonUnitTests.cs:864]: Missing auto format here as well as on the following line, space missing after comma.

**Review by mboud — COMMENTED** [review]
First review here, let me know if I need to change anything regarding the content / form of my comments.

File: nanoFramework.Json/JsonConvert.cs (hunk 1 above)
Comment (Ellerbach) [reply nanoFramework.Json/JsonConvert.cs:508]: Thanks! Found it as well as it broke another test ;-)

File: nanoFramework.Json/JsonConvert.cs (hunk 2 above)
Comment (Ellerbach) [reply nanoFramework.Json/JsonConvert.cs:1147]: good catch, it's not!

File: nanoFramework.Json.Test/JsonUnitTests.cs (hunk 3 above)
Comment (Ellerbach) [reply nanoFramework.Json.Test/JsonUnitTests.cs:878]: ctrl + k + d was still missing on the file. Done. Thx

**Review by josesimoes — APPROVED** [review]

**Pull request merged**